### PR TITLE
Compile crates individually in CI

### PR DESCRIPTION
**Pull request by grtlr**
# Description of change

This is the `mainnet-develop` equivalent of #1222.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

**Review by pvdrz — DISMISSED** [review]

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -67,23 +71,14 @@ jobs:
 
       - uses: Swatinem/rust-cache@v1
 
-      - name: Build
-        uses: actions-rs/cargo@v1
-        with:
-          command: build
-          args: --workspace --all-targets --all-features --release
+      - name: Build crates
+        run: .github/workflows/scripts/for_each_crate.sh "build" "--all-targets --all-features --release"
```
Comment (Alex6323) [inline .github/workflows/ci.yml:75]: Are we just checking here whether the crates successfully build, or do we actually need the created binaries?

Comment (Alex6323) [reply .github/workflows/ci.yml:75]: Because if not, couldn't we drop the `--release` flag?

File: bee-ledger/Cargo.toml [hunk 2]
```diff
@@ -34,6 +34,10 @@ tokio = { version = "1.12.0", default-features = false, features = [ "sync", "fs
 tokio-stream = { version = "0.1.7", default-features = false, optional = true }
 url = { version = "2.2.2", default-features = false, optional = true }
 
+# `udeps` false positives
+[package.metadata.cargo-udeps.ignore]
+normal = [ "time-helper" ]
```
Comment (thibault-martinez) [inline bee-ledger/Cargo.toml]: It should be made optional and put under the `workers` feature instead

File: bee-test/Cargo.toml [hunk 3]
```diff
@@ -21,5 +21,5 @@ rand = { version = "0.8.4", default-features = false }
 
 [dev-dependencies]
 hex = { version = "0.4.3", default-features = false }
-serde = { version = "1.0.130", default-features = false }
-serde_json = { version = "1.0.68", default-features = false, features = [ "alloc" ] }
+serde = { version = "1.0.130", default-features = false, features = [ "alloc" ] }
+serde_json = { version = "1.0.68", default-features = false, features = ["alloc" ] }
```
Comment (thibault-martinez) [inline bee-test/Cargo.toml]:
```suggestion
serde_json = { version = "1.0.68", default-features = false, features = [ "alloc" ] }
```

File: bee-network/bee-autopeering/Cargo.toml [hunk 4]
```diff
@@ -23,18 +23,22 @@ libp2p-core = { version = "0.30.0", default-features = false }
...
+serde = { version = "1.0", default-features = false, features = [ "derive" ] }
 sled = { version = "0.34.7", default-features = false, optional = true }
 thiserror = { version = "1.0.29", default-features = false  }
-tokio = { version = "1.11", default-features = false , features = [ "macros", "rt", "time", "net" ] }
+tokio = { version = "1.11", default-features = false, features = [ "macros", "rt", "time", "net" ] }
 tokio-stream = { version = "0.1", default-features = false  }
 
+# `udeps` false positives
+[package.metadata.cargo-udeps.ignore]
+normal = [ "bincode", "tokio-stream" ]
```
Comment (thibault-martinez) [inline bee-network/bee-autopeering/Cargo.toml]: Are we 100% sure these are false positive and not something that depends on another feature ?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (grtlr) [reply .github/workflows/ci.yml:75]: We test later in this particular job, so by building for release here, we should be able to reuse those binaries.

File: bee-network/bee-autopeering/Cargo.toml (hunk 4 above)
Comment (grtlr) [reply bee-network/bee-autopeering/Cargo.toml]: I fixed `tokio-stream`. `bincode` is required when either `rocksdb` or `sled` are present and my current understanding is that one of them has to be present.

Comment (thibault-martinez) [reply bee-network/bee-autopeering/Cargo.toml]: I think that if you use the `in-memory` feature you don't need rockdb nor sled. So I guess this is not really a false positive and that bincode should go under both features.

File: .github/workflows/scripts/for_each_crate.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+# It expects two inputs:
+#   $1: The cargo command
+#   $2: Any additional flags as string
+
+# Example: `.github/workflows/scripts/for_each_crate.sh build "--all-targets --all-features --release"`
+
+set -e # Abort script on first error.
+
+CRATES=`cargo metadata --format-version 1 --no-deps | jq .packages[].name`
+FLAGS="--all-targets --all-features --release"
```
Comment (thibault-martinez) [inline .github/workflows/scripts/for_each_crate.sh]: Is this unused ?

**Review by thibault-martinez — DISMISSED** [review]

File: .github/workflows/scripts/for_each_crate.sh (hunk 5 above)
Comment (grtlr) [reply .github/workflows/scripts/for_each_crate.sh]:
```suggestion
```

**Review by thibault-martinez — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]

**Review by Adam-Gleave — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Pull request merged**